rcgen/src/lib.rs

```diff
@@ -1352,21 +1339,48 @@ impl CustomExtension {
 
 /// Method to generate key identifiers from public keys.
 ///
-/// This allows choice over methods to generate key identifiers
-/// as specified in RFC 7093 section 2.
+/// Key identifiers should be derived from the public key data. [RFC 7093] defines
+/// three methods to do so using a choice of SHA256 (method 1), SHA384 (method 2), or SHA512
+/// (method 3). In each case the first 160 bits of the hash are used as the key identifier
+/// to match the output length that would be produced were SHA1 used (a legacy option defined
+/// in RFC 5280).
+///
+/// In addition to the RFC 7093 mechanisms, rcgen supports using a pre-specified key identifier.
+/// This can be helpful when working with an existing `Certificate`.
+///
+/// [RFC 7093]: https://www.rfc-editor.org/rfc/rfc7093
 #[derive(Debug, PartialEq, Eq, Hash, Clone)]
 #[non_exhaustive]
 pub enum KeyIdMethod {
-	/// RFC 7093 method 1
+	/// RFC 7093 method 1 - a truncated SHA256 digest.
 	Sha256,
-	/// RFC 7093 method 2
+	/// RFC 7093 method 2 - a truncated SHA384 digest.
 	Sha384,
-	/// RFC 7093 method 3
+	/// RFC 7093 method 3 - a truncated SHA512 digest.
 	Sha512,
-	/// Pre-specified identifier.
+	/// Pre-specified identifier. The exact given value is used as the key identifier.
 	PreSpecified(Vec<u8>),
 }
 
+impl KeyIdMethod {
+	/// Calculates a key identifier for the provided public key using the key ID method.
+	///
+	/// This key identifier is used in the SubjectKeyIdentifier and AuthorityKeyIdentifier
+	/// X.509v3 extensions.
+	pub(crate) fn calculate<K: PublicKeyData>(&self, pub_key: &K) -> Vec<u8> {
```

Comment:
> Nits: I don't love the calculate() name for this. Maybe derive() or digest()?

I went with `derive` since the `PreSpecified` variant could be anything and isn't necessarily a digest at all. 

>  Also I'd suggest eliminating the K type param in favor of pub_key: &impl PublicKeyData.

Done, though this later falls away to `&impl AsRef<[u8]>` to accommodate the change in design where we hold the SPKI and not a `KeyPair` or something implementing `PublicKeyData`.

> Alternatively, maybe it makes sense to rename this to write(), taking a writer argument? Would avoid the to_vec() allocation, at least (and this is private API, anyway).

I tried this but unfortunately because `Certificate::get_key_identifier` exists in the public API we need something we can use without a DER writer. We can't create our own writer and then pull the written bytes back out easily because the bytes are written as a DER encoded octet sequence with a preamble. Trying to write a fn that returns a `Vec<u8>` and one that uses a writer without duplicating code wasn't working out well so I jacked out and left it as-is :laughing: I think the extra allocation will have to stand for now.